tensilelite/Tensile/Source/lib/source/ContractionSolution.cpp

```diff
@@ -775,36 +775,46 @@ namespace Tensile
         }
 
         uint32_t gsu = param.gsu() > 0 ? param.gsu() : sizeMapping.globalSplitU;
-        uint32_t wgm = param.wgm() > 0 ? param.wgm() : sizeMapping.workGroupMapping;
-        if(!internalArgsSupport.wgm)
-            wgm = 0;
-
-        const uint32_t mask         = 0xFF;
-        uint32_t       internalArg0 = mask & gsu;
-        uint32_t       wgShift8     = (mask & wgm) << 8;
-        internalArg0                = internalArg0 | wgShift8;
+        int32_t wgm  = param.wgm() != 0 ? param.wgm() : sizeMapping.workGroupMapping;
+
+        const uint32_t mask16       = 0xFFFF;
+        const uint32_t mask8        = 0xFF;
+        uint32_t       internalArg0 = mask16 & gsu;
+        // FIXME: Temporarily hack, too many custom kernels with different versions
+        if(internalArgsSupport.wgm && sizeMapping.customKernelName != "")
+        {
+            uint32_t       wgShift8     = (mask8 & (uint32_t)wgm) << 8;
+            internalArg0                = internalArg0 | wgShift8;
+        }
+
         // StaggerU
         if(internalArgsSupport.staggerU)
         {
             const uint32_t staggerMask1    = 0x1F00;
             uint32_t       staggerUMapping = (sizeMapping.staggerUMapping << 13);
             uint32_t       staggerUShift   = staggerMask1 & ((sizeMapping.staggerStrideShift) << 8);
-            uint32_t       staggerU        = mask & sizeMapping.staggerU;
+            uint32_t       staggerU        = mask8 & sizeMapping.staggerU;
             staggerU                       = staggerU | staggerUShift;
             staggerU                       = staggerU | staggerUMapping;
             internalArg0                   = internalArg0 | (staggerU << 16);
         }
 
         args.template append<uint32_t>("internalArgs", internalArg0);
+
+        int32_t internalArg1 = 0;
+        if(internalArgsSupport.wgm && sizeMapping.customKernelName == "")
+        {
+            if(wgm == -1)
+                wgm = 1;
+            args.template append<int32_t>("internalArgs1", wgm);
```

Comment:
I think we need to add wgm even if wgm is 0.
Kernel code read this regardless of wgm value.

Reply:
Wgm=0 will still write. The flag is use to check whether wgm is a kernel argument or not.